compiler/src/regular-grammar/semantic.ts

```diff
@@ -75,14 +83,70 @@ class SemanticAnalyzer {
 
     return Object.entries(graph)
       .filter(([, { visited }]) => !visited)
-      .map(([vertex]) => ({ type: 'Warning', message: `${vertex} is unreachable` }));
+      .map(([vertex]) => ({
+        type: 'Warning',
+        message: `${vertex} is unreachable`,
+      }));
+  }
+  // Check cyclic Dependencies from production.
+  private checkCyclicDependencies(): CompileError[] {
+    const graph: Record<string, Set<string>> = {};
+    const visited: Set<string> = new Set();
+    const recStack: Set<string> = new Set();
+    const cycles: string[][] = [];
+
+    // Build the graph
+    this.grammar.rules.forEach(({ lhs, rhs }) => {
+      if (!graph[lhs]) graph[lhs] = new Set();
+      rhs.forEach((token) => {
+        if (token.type[1] === SymbolType.State) {
+          graph[lhs].add(token.value);
+        }
+      });
+    });
+
+    const detectCycle = (node: string, path: string[] = []): boolean => {
+      if (!visited.has(node)) {
+        visited.add(node);
+        recStack.add(node);
+        path.push(node);
+
+        for (const neighbor of graph[node] || []) {
+          if (!visited.has(neighbor) && detectCycle(neighbor, [...path])) {
```

Comment:
If we passed the original path array, all recursive calls would be modifying the same array, which I believe would lead to incorrect results.

Reply:
I still feel it should be possible to pass the same array without a copy. Of course it would require us to handle edge cases of removing the node off when a recursion ends. For now this is fine.